src/proof_production/UnsatCertificateNode.cpp

```diff
@@ -0,0 +1,498 @@
+    }
+
+    _parent = NULL;
+}
+
+void UnsatCertificateNode::setContradiction( Contradiction *contradiction )
+{
+    _contradiction = contradiction;
+}
+
+Contradiction *UnsatCertificateNode::getContradiction() const
+{
+    return _contradiction;
+}
+
+UnsatCertificateNode *UnsatCertificateNode::getParent() const
+{
+    return _parent;
+}
+
+const PiecewiseLinearCaseSplit &UnsatCertificateNode::getSplit() const
+{
+    return _headSplit;
+}
+
+const List<std::shared_ptr<PLCExplanation>> &UnsatCertificateNode::getPLCExplanations() const
+{
+    return _PLCExplanations;
+}
+
+void UnsatCertificateNode::makeLeaf()
+{
+    for ( UnsatCertificateNode *child : _children )
+    {
+        if ( child )
+        {
+            // Clear reference to root tableau so it will not be deleted
+            child->_initialTableau = NULL;
+            delete child;
+            child = NULL;
+        }
+    }
+
+    _children.clear();
+}
+
+void UnsatCertificateNode::passChangesToChildren( UnsatCertificateProblemConstraint *childrenSplitConstraint )
+{
+    for ( auto *child : _children )
+    {
+        child->copyGroundBounds( _groundUpperBounds, _groundLowerBounds );
+        child->_initialTableau = _initialTableau;
+
+        for ( auto &constraint : _problemConstraints )
+        {
+            if ( &constraint != childrenSplitConstraint )
+                child->addProblemConstraint( constraint._type, constraint._constraintVars, constraint._status );
+        }
+
+        // Add the constraint corresponding to head split with correct phase
+        if ( childrenSplitConstraint && childrenSplitConstraint->_type == PiecewiseLinearFunctionType::RELU )
+        {
+            if ( child->_headSplit.getBoundTightenings().front()._type == Tightening::LB || child->_headSplit.getBoundTightenings().back()._type == Tightening::LB  )
+                child->addProblemConstraint( childrenSplitConstraint->_type, childrenSplitConstraint->_constraintVars, PhaseStatus::RELU_PHASE_ACTIVE );
+            else
+                child->addProblemConstraint( childrenSplitConstraint->_type, childrenSplitConstraint->_constraintVars, PhaseStatus::RELU_PHASE_INACTIVE );
+        }
+    }
+}
+
+bool UnsatCertificateNode::certify()
+{
+    // Update ground bounds according to head split
+    for ( auto &tightening : _headSplit.getBoundTightenings() )
+    {
+        auto &temp = tightening._type == Tightening::UB ? _groundUpperBounds : _groundLowerBounds;
+        temp[tightening._variable] = tightening._value;
+    }
+
+    // Certify all PLC bound propagations
+    if ( !certifyAllPLCExplanations( UNSATCertificateUtils::CERTIFICATION_TOLERANCE ) )
+        return false;
+
+    // Save to file if marked
+    if ( _delegationStatus == DelegationStatus::DELEGATE_SAVE )
+        writeLeafToFile();
+
+    // Skip if laf has the SAT solution, or if was marked to delegate
+    if ( _hasSATSolution || _delegationStatus != DelegationStatus::DONT_DELEGATE )
```

Comment:
The comment for this function says:
"      Certifies the tree is indeed a correct proof of unsatisfiability; "

So how can this return true if a leaf has a sat solution? 
Also, the writing to a file that's hiding inside what sounds like it should be a const function is misleading.
Lets talk about this, too